### ISSUE TITLE
BERT - Vocabulary not getting increased post fitting

### ISSUE THREAD

**Issue by Swathygsb** [cdqa-suite/cdQA#226]
I am fine tuning BERT with custom data and found that new words are not added to the oevrall vocabulary and accuracy of the model is only around 10%. Any suggestions ?

**Comment by andrelmfarias** [comment]
Adding new words for task fine-tuning (whether it is MT, Text classification or QA) with BERT is not a good practice, because you won't have proper embeddings for such words. You would need to train the MLM (masked language model) BERT with text that have this vocabulary to have a proper representation for such words.

However, even if there are some domain specific words in your data that are not in BERT vocabulary, the model can still have a satisfactory performance. We test it using financial text (which has a lot of technical terms) as data and it performed well.

In order to understand better what is happening with your custom data, can you please answer the following questions:

- What does 10% of accuracy means here? For QA tasks we use F1-score or Exact Matching (EM). You can check the definitions here: https://arxiv.org/pdf/1606.05250.pdf
- What is the size of your dataset (how many question-answer pairs)?
- Are you using the raw version of the model (i.e., before fine-tune on SQuAD)? Maybe it will perform better if you fine-tune on SQuAD and then on your dataset (especially if your dataset is small)

**Comment by Swathygsb** [comment]
* I have tested around 50 questions and didn't get even 5 of them correctly predicted by the model
* I am fine tuning BERT model with my custom date (around 250 QA pairs) => using .fit method
* For retrieval purpose, I am fitting the retriever with the same 250 Questions again

**Comment by andrelmfarias** [comment]
I am afraid you do not have enough data to fit the model for QA. 

So I advise you to:
- First: fine-tune your model on SQuAD (as per our tutorials), or load one of our models fine-tuned on squad
- Second: do a SECOND fine-tune on your data
- Third: evaluate the performance using f1-score and exact matching, so that it is possible to compare with official metrics

**Comment by Swathygsb** [comment]
I am pretty confused with the prediction part as well. First whether the code finds top_n predictions based tf-idf score and then use bert prediction to get one out of it?
2) Whether the text vectorization will include words from the pre -trained + fitted data as well?

**Comment by andrelmfarias** [comment]
1) The tf-idf retriever find the closest documents indices: 
https://github.com/cdqa-suite/cdQA/blob/a05b4dd667f5021b2c0f66d2ab912b579f733d92/cdqa/pipeline/cdqa_sklearn.py#L155

The BERT reader selects the best answer inside these documents:
https://github.com/cdqa-suite/cdQA/blob/a05b4dd667f5021b2c0f66d2ab912b579f733d92/cdqa/pipeline/cdqa_sklearn.py#L160

2) Which text vectorization are you talking about, tf-idf vectorization or BERT embeddings?
